src/widget/wsearchrelatedtracksmenu.cpp

```diff
@@ -0,0 +1,209 @@
+#include "widget/wsearchrelatedtracksmenu.h"
+
+#include "track/track.h"
+#include "util/qt.h"
+
+namespace {
+
+constexpr double kRelativeBpmRange = 0.06; // +/-6 %
+
+inline int bpmLowerBound(double bpm) {
+    return static_cast<int>(std::floor((1 - kRelativeBpmRange) * bpm));
+}
+
+inline int bpmUpperBound(double bpm) {
+    return static_cast<int>(std::ceil((1 + kRelativeBpmRange) * bpm));
+}
+
+} // namespace
+
+WSearchRelatedTracksMenu::WSearchRelatedTracksMenu(
+        QWidget* parent)
+        : QMenu(tr("Search related Tracks"), parent) {
+}
+
+void WSearchRelatedTracksMenu::addActionsForTrack(
+        const Track& track) {
+    bool addSeparator = false;
+
+    // Musical property actions
+    const auto bpm = track.getBpm();
+    if (bpm > 0) {
+        auto minBpmNumber = QString::number(bpmLowerBound(bpm));
+        auto maxBpmNumber = QString::number(bpmUpperBound(bpm));
+        addSeparator = addSeparatorBeforeAction(addSeparator);
+        addAction(
+                tr("Tempo between %1 and %2 bpm")
+                        .arg(minBpmNumber, maxBpmNumber),
+                [this, minBpmNumber, maxBpmNumber]() {
+                    emit triggerSearch(
+                            QStringLiteral("bpm:>=") +
+                            minBpmNumber +
+                            QStringLiteral(" bpm:<=") +
+                            maxBpmNumber);
+                });
+    }
+    const auto keyText = track.getKeyText();
+    if (!keyText.isEmpty()) {
+        const auto keyTextDisplay =
+                mixxx::escapeTextPropertyWithoutShortcuts(keyText);
+        addSeparator = addSeparatorBeforeAction(addSeparator);
+        addAction(
+                tr("Harmonic keys for \"%1\"")
+                        .arg(keyTextDisplay),
+                [this, keyText]() {
+                    emit triggerSearch(
+                            QStringLiteral("~key:\"") +
+                            keyText + QChar('"'));
+                });
+    }
+    const auto genre = track.getGenre();
+    if (!genre.isEmpty()) {
+        const auto genreDisplay =
+                mixxx::escapeTextPropertyWithoutShortcuts(genre);
+        addSeparator = addSeparatorBeforeAction(addSeparator);
+        addAction(
+                tr("Genre \"%1\"")
+                        .arg(genreDisplay),
+                [this, genre]() {
```

Comment:
Does this use `connect` internally? Asking for a friend.

Reply:
Oh, you are right :see_no_evil: There is an alternative, overloaded (*yikes*) variant that allows to provide the context for the receiver. Since both the action and the menu are part of the same event loop context we can safely omit this redundant context here.

Reply:
The explicit QObject::connect() is problematic, because you explicitly provide a sender pointer which may live in a different context than the receiver.